File: go/pkg/chatgear/listener.go
package chatgear

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/haivivi/giztoy/go/pkg/audio/codec/opus"
	"github.com/haivivi/giztoy/go/pkg/mqtt0"
)

// ErrListenerClosed is returned when Accept is called on a closed Listener.
var ErrListenerClosed = errors.New("chatgear: listener closed")

// =============================================================================
// Listener - Multi-device MQTT server
// =============================================================================

// Listener listens for device connections via MQTT.
// It follows the net.Listener pattern with Accept() blocking until a new device connects.
type Listener struct {
	broker   *mqtt0.Broker
	listener net.Listener
	acceptCh chan *AcceptedPort

	scope   string
	logger  Logger
	timeout time.Duration

	mu     sync.RWMutex
	ports  map[string]*managedPort
	closed bool

	ctx    context.Context
	cancel context.CancelFunc
}

// AcceptedPort represents a newly connected device.
type AcceptedPort struct {
	Port   *ServerPort
	GearID string
}

// managedPort tracks a ServerPort with its activity time.
type managedPort struct {
	port       *ServerPort
	gearID     string
	downlink   *gearDownlink
	lastActive time.Time
}

// gearDownlink implements DownlinkTx for a specific gearID.
type gearDownlink struct {
	listener *Listener
	gearID   string
	scope    string
}

// ListenerConfig contains configuration for the Listener.
type ListenerConfig struct {
	// Addr is the address to listen on (e.g., ":1883").
	Addr string

	// Scope is the topic prefix (e.g., "palr/cn/").
	Scope string

	// Timeout is the inactivity timeout for device connections.
	// Default is 30 seconds.
	Timeout time.Duration

	// Logger is used for logging. If nil, DefaultLogger() is used.
	Logger Logger
}

// ListenMQTT0 creates a new Listener that accepts device connections via MQTT.
//
// Usage:
//
//	ln, err := chatgear.ListenMQTT0(ctx, chatgear.ListenerConfig{Addr: ":1883"})
//	if err != nil {
//	    return err
//	}
//	defer ln.Close()
//
//	for {
//	    accepted, err := ln.Accept()
//	    if err != nil {
//	        break
//	    }
//	    go handleDevice(accepted.Port, accepted.GearID)
//	}
func ListenMQTT0(ctx context.Context, cfg ListenerConfig) (*Listener, error) {
	// Normalize scope
	scope := cfg.Scope
	if scope != "" && !strings.HasSuffix(scope, "/") {
		scope += "/"
	}

	// Set defaults
	logger := cfg.Logger
	if logger == nil {
		logger = DefaultLogger()
	}

	timeout := cfg.Timeout
	if timeout == 0 {
		timeout = 30 * time.Second
	}

	addr := cfg.Addr
	if addr == "" {
		addr = ":1883"
	}

	childCtx, cancel := context.WithCancel(ctx)

	l := &Listener{
		acceptCh: make(chan *AcceptedPort, 32),
		scope:    scope,
		logger:   logger,
		timeout:  timeout,
		ports:    make(map[string]*managedPort),
		ctx:      childCtx,
		cancel:   cancel,
	}

	// Create broker with wildcard handler
	l.broker = &mqtt0.Broker{
		Handler: mqtt0.HandlerFunc(func(clientID string, msg *mqtt0.Message) {
			l.handleMessage(msg.Topic, msg.Payload)
		}),
	}

	// Start listener
	ln, err := mqtt0.Listen("tcp", addr, nil)
	if err != nil {
		cancel()
		return nil, fmt.Errorf("chatgear/listener: listen: %w", err)
	}
	l.listener = ln

	// Start broker serve loop
	go func() {
		if err := l.broker.Serve(ln); err != nil {
			logger.ErrorPrintf("broker serve error: %v", err)
		}
	}()

	// Start timeout checker
	go l.timeoutChecker()

	// Handle context cancellation
	go func() {
		<-childCtx.Done()
		l.closeAll()
	}()

	logger.InfoPrintf("listener started on %s", addr)

	return l, nil
}

// Accept blocks until a new device connects.
// Returns the ServerPort and GearID for the new device.
func (l *Listener) Accept() (*AcceptedPort, error) {
	select {
	case accepted, ok := <-l.acceptCh:
		if !ok {
			return nil, ErrListenerClosed
		}
		return accepted, nil
	case <-l.ctx.Done():
		return nil, l.ctx.Err()
	}
}

// Close closes the Listener and all managed ServerPorts.
func (l *Listener) Close() error {
	l.mu.Lock()
	if l.closed {
		l.mu.Unlock()
		return nil
	}
	l.closed = true
	l.mu.Unlock()

	l.cancel()
	return nil
}

// Addr returns the listener address.
func (l *Listener) Addr() string {
	if l.listener != nil {
		return l.listener.Addr().String()
	}
	return ""
}

// =============================================================================
// Internal methods
// =============================================================================

// Topic patterns for parsing gearID
var (
	topicAudioPattern = regexp.MustCompile(`^(.*)device/([^/]+)/input_audio_stream$`)
	topicStatePattern = regexp.MustCompile(`^(.*)device/([^/]+)/state$`)
	topicStatsPattern = regexp.MustCompile(`^(.*)device/([^/]+)/stats$`)
)

// handleMessage routes incoming MQTT messages to appropriate ServerPorts.
func (l *Listener) handleMessage(topic string, payload []byte) {
	l.logger.DebugPrintf("handleMessage: topic=%s len=%d", topic, len(payload))

	// Parse topic to extract gearID and message type
	var gearID string
	var msgType string

	if matches := topicAudioPattern.FindStringSubmatch(topic); matches != nil {
		gearID = matches[2]
		msgType = "audio"
	} else if matches := topicStatePattern.FindStringSubmatch(topic); matches != nil {
		gearID = matches[2]
		msgType = "state"
	} else if matches := topicStatsPattern.FindStringSubmatch(topic); matches != nil {
		gearID = matches[2]
		msgType = "stats"
	} else {
		// Unknown topic - log for debugging
		l.logger.DebugPrintf("unknown topic: %s", topic)
		return
	}

	// Get or create port
	mp := l.getOrCreatePort(gearID)
	if mp == nil {
		return // listener closed
	}

	// Update last active time
	l.mu.Lock()
	mp.lastActive = time.Now()
	l.mu.Unlock()

	// Route message to port
	switch msgType {
	case "audio":
		frame, t, ok := unstampFrame(payload)
		if !ok {
			l.logger.WarnPrintf("invalid stamped frame from %s", gearID)
			return
		}
		l.logger.DebugPrintf("RX audio from %s: len=%d ts=%v", gearID, len(frame), t.Format("15:04:05.000"))
		mp.port.HandleAudio(&StampedOpusFrame{Timestamp: t, Frame: frame})

	case "state":
		l.logger.InfoPrintf("RX state from %s: %s", gearID, string(payload))
		var evt StateEvent
		if err := json.Unmarshal(payload, &evt); err != nil {
			l.logger.WarnPrintf("failed to unmarshal state from %s: %v", gearID, err)
			return
		}
		mp.port.HandleState(&evt)

		// Check for shutdown/sleep state
		if evt.State == StateShuttingDown || evt.State == StateSleeping {
			l.releasePort(gearID)
		}

	case "stats":
		l.logger.InfoPrintf("RX stats from %s: %s", gearID, string(payload))
		var evt StatsEvent
		if err := json.Unmarshal(payload, &evt); err != nil {
			l.logger.WarnPrintf("failed to unmarshal stats from %s: %v", gearID, err)
			return
		}
		mp.port.HandleStats(&evt)
	}
}

// getOrCreatePort returns an existing port or creates a new one.
func (l *Listener) getOrCreatePort(gearID string) *managedPort {
	l.mu.Lock()
	defer l.mu.Unlock()

	if l.closed {
		return nil
	}

	// Return existing port
	if mp, exists := l.ports[gearID]; exists {
		return mp
	}

	// Create new port
	port := NewServerPort()

	// Create downlink for this gearID
	downlink := &gearDownlink{
		listener: l,
		gearID:   gearID,
		scope:    l.scope,
	}

	// Start WriteTo goroutine
	go port.WriteTo(downlink)

	mp := &managedPort{
		port:       port,
		gearID:     gearID,
		downlink:   downlink,
		lastActive: time.Now(),
	}
	l.ports[gearID] = mp

	l.logger.InfoPrintf("new device connected: %s", gearID)

	// Send to accept channel (non-blocking)
	select {
	case l.acceptCh <- &AcceptedPort{Port: port, GearID: gearID}:
	default:
		l.logger.WarnPrintf("accept channel full, dropping new device %s", gearID)
	}

	return mp
}

// releasePort closes and removes a port.
func (l *Listener) releasePort(gearID string) {
	l.mu.Lock()
	mp, exists := l.ports[gearID]
	if exists {
		delete(l.ports, gearID)
	}
	l.mu.Unlock()

	if exists && mp.port != nil {
		l.logger.InfoPrintf("releasing device: %s", gearID)
		mp.port.Close()
	}
}

// timeoutChecker periodically checks for inactive ports.
func (l *Listener) timeoutChecker() {
	ticker := time.NewTicker(10 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-l.ctx.Done():
			return
		case <-ticker.C:
			l.checkTimeouts()
		}
	}
}

// checkTimeouts checks for and releases inactive ports.
func (l *Listener) checkTimeouts() {
	l.mu.Lock()
	now := time.Now()
	var toRelease []string
	for gearID, mp := range l.ports {
		if now.Sub(mp.lastActive) > l.timeout {
			toRelease = append(toRelease, gearID)
		}
	}
	l.mu.Unlock()

	for _, gearID := range toRelease {
		l.logger.InfoPrintf("device timeout: %s", gearID)
		l.releasePort(gearID)
	}
}

// closeAll closes the listener and all ports.
func (l *Listener) closeAll() {
	l.mu.Lock()
	ports := make([]*managedPort, 0, len(l.ports))
	for _, mp := range l.ports {
		ports = append(ports, mp)
	}
	l.ports = nil
	l.mu.Unlock()

	// Close all ports
	for _, mp := range ports {
		if mp.port != nil {
			mp.port.Close()
		}
	}

	// Close accept channel
	close(l.acceptCh)

	// Close broker and listener
	if l.listener != nil {
		l.listener.Close()
	}
	if l.broker != nil {
		l.broker.Close()
	}
}

// =============================================================================
// gearDownlink - DownlinkTx implementation for a specific gearID
// =============================================================================

func (d *gearDownlink) SendOpusFrame(timestamp time.Time, frame opus.Frame) error {
	topic := fmt.Sprintf("%sdevice/%s/output_audio_stream", d.scope, d.gearID)
	stamped := stampFrame(frame, timestamp)
	d.listener.logger.DebugPrintf("TX audio to %s: len=%d ts=%v", d.gearID, len(frame), timestamp.Format("15:04:05.000"))
	return d.listener.broker.Publish(d.listener.ctx, topic, stamped)
}

func (d *gearDownlink) IssueCommand(cmd Command, t time.Time) error {
	topic := fmt.Sprintf("%sdevice/%s/command", d.scope, d.gearID)
	evt := NewCommandEvent(cmd, t)
	data, err := json.Marshal(evt)
	if err != nil {
		return err
	}
	d.listener.logger.InfoPrintf("TX command to %s: %s", d.gearID, string(data))
	return d.listener.broker.Publish(d.listener.ctx, topic, data)
}

func (d *gearDownlink) Close() error {
	// No-op: the listener manages the lifecycle
	return nil
}

// Compile-time interface assertion
var _ DownlinkTx = (*gearDownlink)(nil)

File: rust/chatgear/src/listener.rs
//! Device discovery and connection acceptance via MQTT.
//!
//! The `Listener` follows the `net::TcpListener` pattern: it starts an embedded
//! MQTT broker and yields `AcceptedPort` instances when new devices connect.
//!
//! # Example
//!
//! ```ignore
//! let listener = listen_mqtt0(ListenerConfig {
//!     addr: "127.0.0.1:1883".to_string(),
//!     scope: "palr/cn".to_string(),
//!     ..Default::default()
//! }, logger).await?;
//!
//! loop {
//!     let accepted = listener.accept().await?;
//!     tokio::spawn(handle_device(accepted.gear_id, accepted.port));
//! }
//! ```

use crate::{
    ConnError, GearStateEvent, GearStatsEvent, StampedOpusFrame,
    conn_mqtt::unstamp_frame,
    logger::Logger,
};
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::{mpsc, Mutex, RwLock};
use tokio_util::sync::CancellationToken;

/// Configuration for the Listener.
#[derive(Debug, Clone)]
pub struct ListenerConfig {
    /// Address to listen on (e.g., "127.0.0.1:1883").
    pub addr: String,
    /// Topic prefix (e.g., "palr/cn").
    pub scope: String,
    /// Inactivity timeout for device connections. Default: 30s.
    pub timeout: Duration,
}

impl Default for ListenerConfig {
    fn default() -> Self {
        Self {
            addr: "127.0.0.1:1883".to_string(),
            scope: String::new(),
            timeout: Duration::from_secs(30),
        }
    }
}

/// A newly connected device.
#[derive(Debug)]
pub struct AcceptedPort {
    /// Device identifier.
    pub gear_id: String,
    /// Receiver for opus frames from this device.
    pub opus_rx: mpsc::Receiver<StampedOpusFrame>,
    /// Receiver for state events from this device.
    pub state_rx: mpsc::Receiver<GearStateEvent>,
    /// Receiver for stats events from this device.
    pub stats_rx: mpsc::Receiver<GearStatsEvent>,
}

/// Tracks a managed device port.
struct ManagedPort {
    gear_id: String,
    opus_tx: mpsc::Sender<StampedOpusFrame>,
    state_tx: mpsc::Sender<GearStateEvent>,
    stats_tx: mpsc::Sender<GearStatsEvent>,
    last_active: Instant,
}

/// Listener for device connections via MQTT.
///
/// Starts an embedded MQTT broker and routes incoming device messages
/// to per-device channels. New devices are announced via `accept()`.
pub struct Listener {
    accept_rx: Mutex<mpsc::Receiver<AcceptedPort>>,
    cancel: CancellationToken,
    logger: Arc<dyn Logger>,
}

impl Listener {
    /// Blocks until a new device connects.
    pub async fn accept(&self) -> Result<AcceptedPort, ConnError> {
        let mut rx = self.accept_rx.lock().await;
        match rx.recv().await {
            Some(port) => Ok(port),
            None => Err(ConnError::Closed),
        }
    }

    /// Closes the listener.
    pub fn close(&self) {
        self.cancel.cancel();
    }
}

/// Creates a new Listener that accepts device connections via MQTT.
///
/// Starts an embedded MQTT broker on the given address. Devices that publish
/// to `{scope}device/{gear_id}/input_audio_stream` are automatically discovered
/// and their data routed to the `AcceptedPort` returned by `accept()`.
pub async fn listen_mqtt0(
    cfg: ListenerConfig,
    log: Arc<dyn Logger>,
) -> Result<Listener, ConnError> {
    let scope = if cfg.scope.is_empty() || cfg.scope.ends_with('/') {
        cfg.scope.clone()
    } else {
        format!("{}/", cfg.scope)
    };

    let timeout = if cfg.timeout == Duration::ZERO {
        Duration::from_secs(30)
    } else {
        cfg.timeout
    };

    let (accept_tx, accept_rx) = mpsc::channel(32);
    let cancel = CancellationToken::new();
    let ports: Arc<RwLock<HashMap<String, ManagedPort>>> = Arc::new(RwLock::new(HashMap::new()));

    // Create broker with message handler
    let handler_ports = ports.clone();
    let handler_scope = scope.clone();
    let handler_accept_tx = accept_tx.clone();
    let handler_log = log.clone();
    let handler_timeout = timeout;

    let broker_cfg = giztoy_mqtt0::BrokerConfig::new(&cfg.addr).sys_events(false);
    let broker = giztoy_mqtt0::Broker::builder(broker_cfg)
        .handler(MessageHandler {
            ports: handler_ports.clone(),
            scope: handler_scope.clone(),
            accept_tx: handler_accept_tx,
            logger: handler_log.clone(),
        })
        .build();

    // Start broker
    let broker_cancel = cancel.clone();
    let broker_log = log.clone();
    tokio::spawn(async move {
        tokio::select! {
            result = broker.serve() => {
                if let Err(e) = result {
                    broker_log.error(&format!("broker serve error: {}", e));
                }
            }
            _ = broker_cancel.cancelled() => {
                broker_log.info("listener broker shutting down");
            }
        }
    });

    // Start timeout checker
    let timeout_ports = ports.clone();
    let timeout_cancel = cancel.clone();
    let timeout_log = log.clone();
    tokio::spawn(async move {
        loop {
            tokio::select! {
                _ = timeout_cancel.cancelled() => return,
                _ = tokio::time::sleep(Duration::from_secs(5)) => {
                    let mut ports = timeout_ports.write().await;
                    let now = Instant::now();
                    ports.retain(|gear_id, port| {
                        if now.duration_since(port.last_active) > handler_timeout {
                            timeout_log.info(&format!("device {} timed out", gear_id));
                            false
                        } else {
                            true
                        }
                    });
                }
            }
        }
    });

    log.info(&format!("listener started on {}", cfg.addr));

    Ok(Listener {
        accept_rx: Mutex::new(accept_rx),
        cancel,
        logger: log,
    })
}

/// MQTT message handler that routes messages to per-device channels.
struct MessageHandler {
    ports: Arc<RwLock<HashMap<String, ManagedPort>>>,
    scope: String,
    accept_tx: mpsc::Sender<AcceptedPort>,
    logger: Arc<dyn Logger>,
}

impl giztoy_mqtt0::Handler for MessageHandler {
    fn handle(&self, _client_id: &str, msg: &giztoy_mqtt0::Message) {
        let topic = &msg.topic;
        let payload = &msg.payload;

        // Parse topic: {scope}device/{gear_id}/{type}
        let prefix = format!("{}device/", self.scope);
        let rest = match topic.strip_prefix(&prefix) {
            Some(r) => r,
            None => return,
        };

        let (gear_id, msg_type) = match rest.find('/') {
            Some(idx) => (&rest[..idx], &rest[idx + 1..]),
            None => return,
        };

        if gear_id.is_empty() {
            return;
        }

        // Use a blocking approach for the async lock since Handler::handle is sync
        let ports = self.ports.clone();
        let accept_tx = self.accept_tx.clone();
        let gear_id = gear_id.to_string();
        let msg_type = msg_type.to_string();
        let payload = payload.to_vec();
        let logger = self.logger.clone();

        tokio::spawn(async move {
            let mut ports_guard = ports.write().await;

            // Get or create port for this gear
            if !ports_guard.contains_key(&gear_id) {
                let (opus_tx, opus_rx) = mpsc::channel(1024);
                let (state_tx, state_rx) = mpsc::channel(32);
                let (stats_tx, stats_rx) = mpsc::channel(32);

                ports_guard.insert(gear_id.clone(), ManagedPort {
                    gear_id: gear_id.clone(),
                    opus_tx,
                    state_tx,
                    stats_tx,
                    last_active: Instant::now(),
                });

                // Notify listener of new device
                let _ = accept_tx.try_send(AcceptedPort {
                    gear_id: gear_id.clone(),
                    opus_rx,
                    state_rx,
                    stats_rx,
                });

                logger.info(&format!("new device connected: {}", gear_id));
            }

            if let Some(port) = ports_guard.get_mut(&gear_id) {
                port.last_active = Instant::now();

                match msg_type.as_str() {
                    "input_audio_stream" => {
                        if let Some(frame) = unstamp_frame(&payload) {
                            let _ = port.opus_tx.try_send(frame);
                        }
                    }
                    "state" => {
                        if let Ok(evt) = serde_json::from_slice::<GearStateEvent>(&payload) {
                            let _ = port.state_tx.try_send(evt);
                        }
                    }
                    "stats" => {
                        if let Ok(evt) = serde_json::from_slice::<GearStatsEvent>(&payload) {
                            let _ = port.stats_tx.try_send(evt);
                        }
                    }
                    _ => {}
                }
            }
        });
    }
}
